src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
     // - eventArgs: the PasteFromClipboard event sent from the TermControl
     fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/,
                                                              const PasteFromClipboardEventArgs eventArgs)
+    try
     {
-        const auto data = Clipboard::GetContent();
-
-        // This will switch the execution of the function to a background (not
-        // UI) thread. This is IMPORTANT, because the getting the clipboard data
-        // will crash on the UI thread, because the main thread is a STA.
-        co_await winrt::resume_background();
+        // Counterintuitively pasting the clipboard should block the UI thread, because that's
+        // actually the expectation of users: If you type something, paste and then type something again,
+        // text should be written in exactly that order, no matter how loaded the CPU is. See GH#15315.
+        assert(Dispatcher().HasThreadAccess());
 
-        try
+        if (!IsClipboardFormatAvailable(CF_UNICODETEXT))
```

Comment:
what about `CF_OEMTEXT` and `CF_TEXT`?

I would guess that `StandardDataFormats::Text` handles that for us. Now we need to do it ourselves.

After fixing this, is it still as much faster?

Reply:
> what about `CF_OEMTEXT` and `CF_TEXT`?

The 1992 API is actually way more 2023 than one might think. 😄 It has "synthesized types", so that if one application writes a `CF_OEMTEXT` you can still receive it as a `CV_UNICODE_TEXT` if that's what you're asking for. See here: https://learn.microsoft.com/en-us/windows/win32/dataxchg/clipboard-formats#synthesized-clipboard-formats

> After fixing this, is it still as much faster?

No fixing needed, but even if we did, it'll always be faster. The only "expensive" part of the old API is `OpenClipboard` which acquires an mutex in the kernel. That takes roughly 10us on my PC. All other calls are practically copying data straight out of it for free. The WinRT API on the other hand does 1 RPC call for anything you do on the `DataPackageView` object (3500us per call on my PC). So just the `.Contains(StandardDataFormats.Text)` alone is worth _thousands_ of win32 clipboard API calls.

> I would guess that `StandardDataFormats::Text` handles that for us. Now we need to do it ourselves.

Actually the "ClipboardServer" process is an entirely standalone thing that isn't built on top of the win32 API at all. Instead, it's basically... oh well... oh no...
Records indicate that ClipboardServer is actually just the _exact same thing as the win32 API_! It also runs as a single, global instance and it also is nothing but a `SRWLOCK` that protects a global, single agile instance of a `DataPackageView` that is then lent out to anyone calling `Clipboard::GetContent`. Software engineering is a circle. 😄

So yeah... it's actually just a mutex around a `DataPackageView`. I _think_ it stays in sync with the win32 side of things via the Windows Notification Facility (WNF), which, if true, would also be an absolute banger. 😄

Reply:
@KalleOlaviNiemitalo Thanks! That's a good point. I think adding the `IsClipboardFormatAvailable` is a bit of a premature optimization anyways, I think. Usually when you right click into Windows Terminal you got something to paste in your clipboard already. 🤔

Reply:
Actually, nevermind... It seems to work just fine. I've tested it using this application:
```cpp
#define NOMINMAX
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>

int main() {
    OpenClipboard(nullptr);
    EmptyClipboard();

    const auto mem = GlobalAlloc(GMEM_MOVEABLE, 4);
    memcpy(GlobalLock(mem), "foo", 4);
    GlobalUnlock(mem);

    SetClipboardData(CF_TEXT, mem);
    CloseClipboard();
    return 0;
}
```

I can paste "foo" just fine in WT afterwards. (Also, I'm surprised that `OpenClipboard(nullptr)` works! That's pretty neat.)

Reply:
If you'd like, I think we should continue talking about these things in [the comment below](https://github.com/microsoft/terminal/pull/15360#discussion_r1194577685). Regarding `ClipboardServer.dll` however I'd like to already bring up that it is in fact involved when calling `Clipboard::GetContents`. Its actual runtime is so low though (compared to COM and OLE) that I couldn't figure out what exactly it's doing unfortunately. Here's the only traces I got for the brief ~3s test runtime:
```
Name                                                                                   Self, %    Stack, %    Module Name
Microsoft::WRL::FtmBase::MarshalInterface                                              ·          1.22        C:\Windows\System32\ClipboardServer.dll
Microsoft::WRL::Details::RuntimeClassImpl<...,IClipboardServer,...>::QueryInterface    0.07       0.09        C:\Windows\System32\ClipboardServer.dll
Microsoft::WRL::FtmBase::GetMarshalSizeMax                                             ·          0.04        C:\Windows\System32\ClipboardServer.dll
InlineIsEqualGUID                                                                      0.02       0.02        C:\Windows\System32\ClipboardServer.dll
ClipboardServer::GetBroker                                                             0.02       0.02        C:\Windows\System32\ClipboardServer.dll
```